### PR TITLE
CP-37692: Introduce RBAC role for client-auth sessions

### PR DESCRIPTION
**Pull request by robhoes**

### CODE REVIEW

File: ocaml/idl/datamodel_roles.ml [hunk 1]
```diff
@@ -6,9 +6,12 @@ let role_vm_power_admin = "vm-power-admin"
 let role_vm_admin = "vm-admin"
 let role_vm_operator = "vm-operator"
 let role_read_only = "read-only"
+let role_client_cert = "client-cert"
```
Comment (edwintorok) [inline ocaml/idl/datamodel_roles.ml:9]: I think these are roughly sorted by level of privilege, so this should go near pool operator here.

File: ocaml/idl/datamodel_roles.ml [hunk 2]
```diff
@@ -22,6 +25,7 @@ let role_description = [
   role_vm_admin,"The VM Administrator role can manage VMs and templates";
   role_vm_operator,"The VM Operator role can use VMs and interact with VM consoles";
   role_read_only,"The Read-Only role can log in with basic read-only access";
+  role_client_cert,"The Client Certificate role has access to funtionality granted to those who authenticate using a client certificate"
```
Comment (edwintorok) [inline ocaml/idl/datamodel_roles.ml:28]: as above, I think the list should be kept sorted.

File: ocaml/xapi/xapi_session.ml [hunk 3]
```diff
@@ -711,15 +711,35 @@ let login_with_password ~__context ~uname ~pwd ~version ~originator =
   let pwd = Bytes.of_string pwd in
   wipe_params_after_fn [pwd] (fun () ->
       (* !!! Do something with the version number *)
-      if Context.preauth ~__context then
+      match Context.preauth ~__context with
+      | Some socket_kind when socket_kind = `root ->
```
Comment (edwintorok) [inline ocaml/xapi/xapi_session.ml:715]: the `when` is redundant here you should be able to write:`Some 'root`, and that will also properly check exhaustiveness.

File: ocaml/xapi/xapi_session.ml [hunk 4]
```diff
@@ -711,15 +711,35 @@ let login_with_password ~__context ~uname ~pwd ~version ~originator =
...
         (* in this case, the context origin of this login request is a unix socket bound locally to a filename *)
         (* we trust requests from local unix filename sockets, so no need to authenticate them before login *)
         login_no_password_common ~__context ~uname:(Some uname) ~originator
           ~host:(Helpers.get_localhost ~__context)
           ~pool:false ~is_local_superuser:true ~subject:Ref.null
           ~auth_user_sid:"" ~auth_user_name:uname ~rbac_permissions:[]
           ~db_ref:None
-      else
+      | Some socket_kind when socket_kind = `client_cert ->
```
Comment (edwintorok) [inline ocaml/xapi/xapi_session.ml:723]: as above

File: ocaml/xapi/xapi_session.ml [hunk 5]
```diff
@@ -711,15 +711,35 @@ let login_with_password ~__context ~uname ~pwd ~version ~originator =
...
+                 (Api_errors.internal_error, [Datamodel_roles.role_client_cert ^ " role not found"])
+              )
+        in
+        let rbac_permissions = Xapi_role.get_permissions_name_label ~__context ~self:role in
+        login_no_password_common ~__context ~uname:(Some uname) ~originator
+          ~host:(Helpers.get_localhost ~__context)
+          ~pool:false ~is_local_superuser:false ~subject:Ref.null
+          ~auth_user_sid:"" ~auth_user_name:uname ~rbac_permissions
+          ~db_ref:None
+      | Some _ | None ->
```
Comment (edwintorok) [inline ocaml/xapi/xapi_session.ml]: You should remove the `Some` and let OCaml's exhaustiveness analysis prove that you handled both kinds.

File: ocaml/idl/datamodel_pool.ml [hunk 6]
```diff
@@ -712,7 +712,7 @@ open Datamodel_types
         Ref _pool, "self", "The pool";
         Set (Ref _repository), "value", "The set of repositories to be enabled"
       ]
-      ~allowed_roles:_R_POOL_ADMIN
+      ~allowed_roles:_R_POOL_OP
```
Comment (edwintorok) [inline ocaml/idl/datamodel_pool.ml:715]: Do we have a baked-in gpg key and we only trust RPMs signed by that gpg key when using the repository?
(I think this is what happens with updates, but I'm not sure what happens with a Yum repository, do we just trust all RPMs inside it by default, or only those signed by us?) 
If not then pool operator could escalate itself to pool admin by creating a Yum repository itself, replacing XAPI rpm with one that disables all authentication.

Comment (robhoes) [reply ocaml/idl/datamodel_pool.ml:715]: Yes indeed, that is the reason: you can only install signed RPMs, and a pool-operator cannot change which key is accepted.

File: ocaml/idl/datamodel_roles.ml (hunk 1 above)
Comment (robhoes) [reply ocaml/idl/datamodel_roles.ml:9]: I wanted to keep the original set of roles together.

File: ocaml/idl/datamodel_pool.ml (hunk 6 above)
Comment (edwintorok) [reply ocaml/idl/datamodel_pool.ml:715]: I assume we (will) have a test for that.

**Review by edwintorok — APPROVED** [review]

**Review by rwdrich — APPROVED** [review]

**Pull request merged**